rust/user-info-fetcher/src/backend/entra.rs

```diff
@@ -0,0 +1,267 @@
+    #[serde(default)]
+    attributes: HashMap<String, serde_json::Value>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct GroupMembership {
+    display_name: String,
+}
+
+struct EntraEndpoint {
+    hostname: HostName,
+    port: u16,
+    tenant_id: String,
+    protocol: String,
+}
+
+pub(crate) async fn get_user_info(
+    req: &UserInfoRequest,
+    http: &reqwest::Client,
+    credentials: &Credentials,
+    config: &v1alpha1::EntraBackend,
+) -> Result<UserInfo, Error> {
+    let v1alpha1::EntraBackend {
+        client_credentials_secret: _,
+        hostname,
+        port,
+        tenant_id,
+        tls,
+    } = config;
+
+    let entra_endpoint = EntraEndpoint::new(hostname.clone(), port.clone(), tenant_id.clone(), tls);
+    let token_url = entra_endpoint.oauth2_token();
+
+    let authn = send_json_request::<OAuthResponse>(http.post(token_url).form(&[
+        ("client_id", credentials.client_id.as_str()),
+        ("client_secret", credentials.client_secret.as_str()),
+        ("scope", "https://graph.microsoft.com/.default"),
+        ("grant_type", "client_credentials"),
+    ]))
+    .await
+    .context(AccessTokenSnafu)?;
+
+    let user_info = match req {
+        UserInfoRequest::UserInfoRequestById(req) => {
+            let user_id = req.id.clone();
+            send_json_request::<UserMetadata>(
+                http.get(entra_endpoint.user_info(&user_id))
+                    .bearer_auth(&authn.access_token),
+            )
+            .await
+            .context(UserNotFoundByIdSnafu { user_id })?
+        }
+        UserInfoRequest::UserInfoRequestByName(req) => {
+            let username = &req.username;
+            send_json_request::<UserMetadata>(
+                http.get(entra_endpoint.user_info(&username))
+                    .bearer_auth(&authn.access_token),
+            )
+            .await
+            .context(SearchForUserSnafu)?
+        }
+    };
+
+    let groups = send_json_request::<Vec<GroupMembership>>(
+        http.get(entra_endpoint.group_info(&user_info.id))
+            .bearer_auth(&authn.access_token),
+    )
+    .await
+    .context(RequestUserGroupsSnafu {
+        username: user_info.user_principal_name.clone(),
+        user_id: user_info.id.clone(),
+    })?;
+
+    Ok(UserInfo {
+        id: Some(user_info.id),
+        username: Some(user_info.user_principal_name),
+        groups: groups.into_iter().map(|g| g.display_name).collect(),
+        custom_attributes: user_info.attributes,
+    })
+}
+
+impl EntraEndpoint {
+    pub fn new(hostname: HostName, port: u16, tenant_id: String, tls: &TlsClientDetails) -> Self {
+        Self {
+            hostname,
+            port,
+            tenant_id,
+            protocol: if tls.uses_tls() {
+                "https".to_string()
+            } else {
+                "http".to_string()
+            },
+        }
+    }
+
+    pub fn oauth2_token(&self) -> String {
+        format!(
```

Comment:
nit: We could use https://docs.rs/url/latest/url/struct.Url.html for all URL-related computation and return an URL
Watch out, https://docs.rs/url/latest/url/struct.Url.html#method.join sucks :D